src/puzzles/day_8.rs

```diff
@@ -0,0 +1,136 @@
+
+    #[derive(Debug, Clone)]
+    pub enum Instruction {
+        Accumulate(i32),
+        Jump(i32),
+        NoOperation(i32),
+    }
+
+    impl Instruction {
+        pub fn from(instruction: &str) -> Result<Instruction, &str> {
+            match &instruction[..3] {
+                "acc" => match instruction[4..].parse::<i32>() {
+                    Ok(number) => return Ok(Instruction::Accumulate(number)),
+                    Err(_) => {
+                        return Err("Could not parse number from accumulate instruction.")
+                    }
+                },
+                "jmp" => match instruction[4..].parse::<i32>() {
+                    Ok(offset) => return Ok(Instruction::Jump(offset)),
+                    Err(_) => return Err("Could not parse offset from jump instruction."),
+                },
+                "nop" => match instruction[4..].parse::<i32>() {
+                    Ok(number) => return Ok(Instruction::NoOperation(number)),
+                    Err(_) => {
+                        return Err("Could not parse number from no operation instruction.")
+                    }
+                },
+                _ => return Err("Unknown instruction. Must be acc, jmp or nop"),
+            }
+        }
+    }
+
+    #[derive(Debug, PartialEq)]
+    pub enum Termination {
+        Normal,
+        InfiniteLoop
+    }
+
+    #[derive(Debug)]
+    pub struct Computer {
+        accumulator: i32,
+        instructions: Vec<Instruction>,
+        instruction_pointer: usize,
+    }
+
+    impl Computer {
+        pub fn new() -> Computer {
+            Computer {
+                accumulator: 0,
+                instructions: Vec::new(),
+                instruction_pointer: 0,
+            }
+        }
+
+        pub fn from(instructions: Vec<Instruction>) -> Computer {
+            Computer {
+                accumulator: 0,
+                instructions,
+                instruction_pointer: 0,
+            }
+        }
+
+        pub fn load(&mut self, instructions: Vec<Instruction>) {
+            self.accumulator = 0;
+            self.instructions = instructions;
+            self.instruction_pointer = 0;
+        }
+
+        fn step(&mut self) {
+            match self.instructions.get(self.instruction_pointer).expect("puzzles::day_8::computation::Computer tried to access invalid memory (instruction pointer out of bounds)") {
+                Instruction::Accumulate(number) => {
+                    self.accumulator += number;
+                    self.instruction_pointer += 1;
+                },
+                Instruction::Jump(offset) => self.instruction_pointer = usize::try_from(self.instruction_pointer as i32 + offset).expect("Invalid jump. New instruction pointer out of bounds!"),
+                Instruction::NoOperation(_) => self.instruction_pointer += 1
+            }
+        }
+
+        /// Execute Instructions until hitting a loop.
+        pub fn run(&mut self) -> Termination {
```

Comment:
Instead of returning a termination status consider setting a flag in the computer to indicate getting into an infinite loop, like `computer.termination_status()`. Or even do both...